python/src/array.cpp

```diff
@@ -131,30 +207,42 @@ PyScalarT validate_shape(T list, const std::vector<int>& shape, int idx) {
   }
 
   if (s == 0) {
-    return pyfloat;
+    return float32;
   }
 
-  PyScalarT type = pybool;
+  Dtype type = bool_;
   for (auto l : list) {
-    PyScalarT t;
+    Dtype t = bool_;
     if (py::isinstance<py::list>(l)) {
       t = validate_shape(l.template cast<py::list>(), shape, idx + 1);
     } else if (py::isinstance<py::tuple>(*list.begin())) {
       t = validate_shape(l.template cast<py::tuple>(), shape, idx + 1);
     } else if (py::isinstance<py::bool_>(l)) {
-      t = pybool;
+      t = bool_;
     } else if (py::isinstance<py::int_>(l)) {
-      t = pyint;
+      t = int32;
     } else if (py::isinstance<py::float_>(l)) {
-      t = pyfloat;
+      t = float32;
     } else if (PyComplex_Check(l.ptr())) {
-      t = pycomplex;
+      t = complex64;
+    } else if (py::isinstance<array>(l)) {
+      auto arr = py::cast<array>(l);
+      if (arr.shape().size() + idx + 1 == shape.size() &&
+          std::equal(
+              arr.shape().cbegin(),
+              arr.shape().cend(),
+              shape.cbegin() + idx + 1)) {
+        t = arr.dtype();
+      } else {
+        throw std::invalid_argument(
+            "Initialization encountered non-uniform length.");
+      }
```

Comment:
Tested in [assertRaises here](https://github.com/ml-explore/mlx/blob/ebe6dddc3da74e875d92d96e339ed7a5dccbe528/python/tests/test_array.py#L259-L264).